src/libostree/ostree-bootloader-zipl.c

```diff
@@ -195,73 +199,117 @@ static gboolean
+    access(SECURE_EXECUTION_LUKS_BOOT_KEY, F_OK) == 0);
 }
 
 static gboolean
-_ostree_secure_execution_enable_luks(const gchar *oldramfs,
-                                     const gchar *newramfs,
-                                     GError **error)
+_ostree_secure_execution_append_luks_keys (int initrd_fd,
+                                           GCancellable *cancellable,
+                                           GError **error)
 {
-  const char *const argv[] = {SECURE_EXECUTION_RAMDISK_TOOL, oldramfs, newramfs, NULL};
-  g_autofree gchar *out = NULL;
-  g_autofree gchar *err = NULL;
-  int status = 0;
-  if (!g_spawn_sync (NULL, (char**)argv, NULL, G_SPAWN_SEARCH_PATH,
-                     NULL, NULL, &out, &err, &status, error))
-    return glnx_prefix_error(error, "s390x SE: spawning %s", SECURE_EXECUTION_RAMDISK_TOOL);
-
-  if (!g_spawn_check_exit_status (status, error))
+#ifdef HAVE_LIBARCHIVE
+  // appending cpio gzip archive with LUKS keys
+  g_autoptr(OtAutoArchiveWrite) a = archive_write_new ();
+  if (archive_write_set_format_cpio_newc (a) != 0 ||
+      archive_write_add_filter_gzip (a) != 0 ||
+      archive_write_open_fd(a, initrd_fd) != 0)
+    return glnx_prefix_error (error, "s390x SE: initing cpio: %s", archive_error_string (a));
+
+  const char *files[] = {"/etc", "/etc/luks", SECURE_EXECUTION_LUKS_CONFIG, SECURE_EXECUTION_LUKS_BOOT_KEY, SECURE_EXECUTION_LUKS_ROOT_KEY};
+  for (uint i = 0; i != sizeof (files)/sizeof (char*); ++i)
     {
-      g_printerr("s390x SE: `%s` stdout: %s\n", SECURE_EXECUTION_RAMDISK_TOOL, out);
-      g_printerr("s390x SE: `%s` stderr: %s\n", SECURE_EXECUTION_RAMDISK_TOOL, err);
-      return glnx_prefix_error(error, "s390x SE: `%s` failed", SECURE_EXECUTION_RAMDISK_TOOL);
+      const char *path = files[i];
+      struct stat st;
+      if (stat(path, &st) != 0)
+        glnx_throw_errno_prefix (error, "s390x SE: stat(%s) failed", path);
+
+      struct archive_entry *ae = archive_entry_new ();
+      archive_entry_copy_stat (ae, &st);
+      archive_entry_set_pathname (ae, path);
+      if (archive_write_header (a, ae) != 0)
+          glnx_prefix_error (error, "s390x SE: writting cpio header: %s", archive_error_string (a));
+
+      if (S_ISREG (st.st_mode))
+        {
+          ot_journal_print(LOG_INFO, "s390x SE: appending %s to initrd", path);
+          glnx_autofd int fd = -1;
+          if (!glnx_openat_rdonly (AT_FDCWD, path, TRUE, &fd, error))
+            return glnx_prefix_error (error, "s390x SE: opening %s", path);
+          g_autoptr(GBytes) data = glnx_fd_readall_bytes (fd, cancellable, error);
+          if (!data)
+            return glnx_prefix_error (error, "s390x SE: reading %s", path);
+
+          gsize size = 0;
+          const char *ptr = (const char *) g_bytes_get_data (data, &size);
+          if (archive_write_data (a, ptr, size) != size)
+            return glnx_prefix_error (error, "s390x SE: writting cpio entry: %s", archive_error_string (a));
```

Comment:
Uhm, I think `archive_error_string(a)` only works for the case where we get a `-1`. I don't think we can use that in the case of short writes.
Overall, I think the two error cases should be handled each one separately.